app/src/main/java/com/android/streetworkapp/utils/ParkFilter.kt

```diff
@@ -0,0 +1,153 @@
+package com.android.streetworkapp.utils
+
+import android.util.Log
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.mutableIntStateOf
+import androidx.compose.runtime.snapshots.SnapshotStateList
+import androidx.compose.runtime.toMutableStateList
+import com.android.streetworkapp.model.park.Park
+
+/**
+ * This class add filtering capabilities to a list of parks.
+ *
+ * @param filterSettings The filter settings.
+ */
+class ParkFilter(private val filterSettings: FilterSettings) {
+
+  /**
+   * Filters a park based on the set filter settings.
+   *
+   * @param park The park to filter.
+   * @return True if the park passes the filter, false otherwise.
+   */
+  fun filter(park: Park): Boolean {
+    // Log.d("ParkFilter", "Filtering park ${park.name}")
+
+    return filterRating(park) && filterEventsDensity(park)
+  }
+
+  /**
+   * Filters a park according to the required minimal rating (filter settings).
+   *
+   * @param park The park to filter.
+   * @return True if the park passes the filter, false otherwise.
+   */
+  private fun filterRating(park: Park): Boolean {
+    return park.rating >= filterSettings.minRating.value
+  }
+
+  /**
+   * Filters a park according to its event density (filter settings).
+   *
+   * @param park The park to filter.
+   * @return The filtered list of parks.
+   */
+  private fun filterEventsDensity(park: Park): Boolean {
+    return filterSettings.eventDensity.any { it.isIncluded(park.events.size) }
+  }
+}
+
+/**
+ * Represents the filter settings.
+ *
+ * @property minRating The minimum rating required.
+ * @property eventDensity The required event density.
+ */
+class FilterSettings {
+  val minRating: MutableState<Int> = mutableIntStateOf(DEFAULT_RATING)
+  val eventDensity: SnapshotStateList<EventDensity> = DEFAULT_EVENT_DENSITY.toMutableStateList()
+
+  /** The default filter settings. */
+  companion object {
+    val DEFAULT_RATING = 1
+    val DEFAULT_EVENT_DENSITY = listOf(EventDensity.LOW, EventDensity.MEDIUM, EventDensity.HIGH)
```

Comment:
Actually the default filters are the following : 

- Minimum rating of 1 star (the minimum rating possible to give to a park)
- By default all kinds of event densities are accepted, giving an empty list would mean no park to display.

The sole thing which could be problematic is if when a park is added, its rating is initialized outside the allowed range [1, 5]. But I do not know how it is defined.

Reply:
What i meant is that i think it would be better to make it that your density filters all start as unselected and accept all park densities in such a default state, then if a user clicks on HIGH density for example, then it will show only the high density parks, the other way around which you are currently doing feels a bit confusing to me, this might just be a detail, if you feel the filters are fine as is i'll still approve the review

Reply:
I think that this is a bad idea, because if a user does not choose to select any density, then no park will be displayed.

The current implementation will display all of the parks if the user touches nothing. Then he can opt-out some parks.